imap/caldav_alarm.c

```diff
@@ -2343,3 +2320,92 @@ static int upgradev4(sqldb_t *db)
 
     return r;
 }
+
+struct floating_rock {
+    struct caldav_db *caldavdb;
+    struct mailbox *mailbox;
+    const char *userid;
+    icaltimezone *tz;
+    time_t runtime;
+};
+
+static int floating_cb(void *rock, struct caldav_data *cdata)
+{
+    struct floating_rock *frock = (struct floating_rock *) rock;
+    struct index_record record = {0};
+    icalcomponent *ical = NULL;
+
+    int r = mailbox_find_index_record(frock->mailbox, cdata->dav.imap_uid, &record);
+    if (!r && (ical = record_to_ical(frock->mailbox, &record, NULL))) {
+        if (has_alarms(ical, frock->mailbox, record.uid, NULL)) {
+            time_t nextcheck = process_alarms(mailbox_name(frock->mailbox),
```

Comment:
A faster option here (just to make this action cheap) might be to update all the nextcheck to `now` and then have calalarmd process the alarms - that way it does all the batching and starvation prevention internally, and avoids this potentially processing thousands or millions of events in a single blocking operation.

Reply:
Done